### ISSUE TITLE
Missing Lens Information for Nikon Z6 + FTZ + F-Mount Lens

### ISSUE THREAD

**Issue by hassec** [Exiv2/exiv2#1433]
**Describe the bug**

The problem I encountered is trying to get lens info out of an image file produced with a Nikon Z6 + FTZ Adapter + 70-200mm f2.8/G IF ED VR
```
❯ exiv2 -pt --grep Lens CH0_0140.NEF
Exif.Nikon3.LensType                         Byte        1  D G VR
Exif.Nikon3.Lens                             Rational    4  70-200mm F2.8
Exif.Nikon3.LensFStops                       Undefined   4  6
Exif.Nikon3.LensData                         Undefined 108  48 56 48 49 189 42 43 240 44 217 10 107 129 69 232 21 173 104 130 64 47 23 15 206 202 144 73 47 215  ...
Exif.Photo.LensSpecification                 Rational    4  700/10 2000/10 280/100 280/100
Exif.Photo.LensMake                          Ascii       6  
Exif.Photo.LensModel                         Ascii      65  
Exif.Photo.LensSerialNumber                  Ascii      11  
```

Compared to a picture taken with a native Z mount lens:
```
Exif.Nikon3.LensType                         Byte        1  (0)
Exif.Nikon3.Lens                             Rational    4  24-70mm F4
Exif.Nikon3.LensFStops                       Undefined   4  5
Exif.Nikon3.LensData                         Undefined 108  48 56 48 49 149 240 134 87 99 170 44 233 225 20 130 43 15 46 136 29 237 248 62 191 123 114 164 17 18 ...
Exif.Photo.LensSpecification                 Rational    4  240/10 700/10 400/100 400/100
Exif.Photo.LensMake                          Ascii       6  NIKON
Exif.Photo.LensModel                         Ascii      65  NIKKOR Z 24-70mm f/4 S
```

As you can see it's missing for example `LensMake`, `LensModel`, and `LensSerialNumber` information.

I googled a bit and found [this link](https://exiftool.org/forum/index.php?topic=9709.0) describing a similar problem in exiftool.

This was fixed in exiftool shortly after in [this github commit](https://github.com/exiftool/exiftool/commit/bd34970d642d8bd6191909ac857a075be5468275)

And now it produces:
```
❯ ./exiftool -a -G1 -s "-*lens*" CH0_0140.NEF
[Nikon]         LensType                        : G VR
[Nikon]         Lens                            : 70-200mm f/2.8
[Nikon]         LensFStops                      : 6.00
[Nikon]         LensDataVersion                 : 0801
[Nikon]         LensIDNumber                    : 119
[Nikon]         LensFStops                      : 6.00
[ExifIFD]       LensInfo                        : 70-200mm f/2.8
[ExifIFD]       LensMake                        : 
[ExifIFD]       LensModel                       : 
[ExifIFD]       LensSerialNumber                : 
[Composite]     LensID                          : AF-S VR Zoom-Nikkor 70-200mm f/2.8G IF-ED
[Composite]     LensSpec                        : 70-200mm f/2.8 G VR
```

which still misses the above mentioned fields but adds `LensID` and provides a `LensIDNumber`.

exiv2 fails to produce a `LensIDNumber` and thus doesn't allow me to workaround missing information with a local config file as is described in the man pages.

Note, I tried with the current release, and build the master and 0.27-maintenance but had the same result for all three versions

**To Reproduce**
I've provided a file to @clanmills 

**Expected behavior**
Ideally provide `LensMake` and `LensModel` but at least a `LensIDNumber`

**Desktop (please complete the following information):**
 - OS: Linux
 - Compiler & Version Gcc 10.2
 - Compilation release mode

**Comment by clanmills** [comment]
I think ExifTool knows how to decode LensData and Exiv2 can't decode that.: https://exiv2.org/tags-nikon.html
![screenshot_37](https://user-images.githubusercontent.com/529982/101944142-27b9c380-3be4-11eb-8a08-26c948a0507a.png)
There's a binary record of 108 bytes.  There's a conversation here about it.  Appears to encrypted:  https://exiftool.org/forum/index.php?topic=6139.0


```
567 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance $ exiv2 -g lens/i ~/Downloads/CH0_0174.NEF  
Exif.Nikon3.LensType                         Byte        1  D G VR
Exif.Nikon3.Lens                             Rational    4  70-200mm F2.8
Exif.Nikon3.LensFStops                       Undefined   4  6
Exif.Nikon3.LensData                         Undefined 108  48 56 48 49 254 111 102 54 19 20 199 37 85 24 149 218 234 173 199 3 108 90 202 9 133 205 121 108 20 247 21 110 2 209 219 32 160 91 81 130 238 149 119 148 236 127 77 86 154 25 211 200 248 99 9 234 6 93 239 188 196 7 133 62 50 97 203 112 80 107 193 82 30 37 103 228 156 143 189 38 202 169 195 24 168 115 121 186 54 237 223 12 116 23 245 45 98 241 187 192 0 123 49 34 78 181 87 52
Exif.Photo.LensSpecification                 Rational    4  700/10 2000/10 280/100 280/100
Exif.Photo.LensMake                          Ascii       6  
Exif.Photo.LensModel                         Ascii      65  
Exif.Photo.LensSerialNumber                  Ascii      11  
568 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance $ 
```
I'll study this next week.

**Comment by clanmills** [comment]
I've looked at this and regret that I cannot help.  The main reason is that the data is encrypted and I don't know the algorithm.  There is a Nikon decoded in dcraw.c which is really available on the internet.  When I run you file with that code it outputs:

```

**Comment by hassec** [comment]
Thanks for having a look @clanmills.  
I only see an empty text box above though, is it maybe missing something?

**Comment by clanmills** [comment]
I clicked the "comment" button in error.  You can read about it now!

**Comment by clanmills** [comment]
I see ExifTool references the decryption at: 4) http://www.cybercom.net/~dcoffin/dcraw/.  It' rather odd that my version of that code does not decode this data structure.  Oh well, something's going to have fun with this.

**Comment by clanmills** [comment]
The latest version of Dave's code is here has has the code to decrypt 0x98:  https://github.com/ncruces/dcraw/blob/master/parse.c

**Comment by clanmills** [comment]
**Good News:**

I've added the nikon_decrypt() function from parse.c to the tvisitor.cpp code in my book.  At the moment, it is not successfully decrypting the data because the function expects the data to start "02" and it starts "08".  Puzzling.  Some debugging in parse.c and I'll probably solve that.

As far as Exiv2 is concerned, I am retired and will not be submit code in future.  However, I want the book to deal with everything I know about Exiv2 to help a future maintainer to understand the code (which is non-trivial).  Currently in the book, I say:

> Metadata is occasionally compressed (PNG), encrypted (Nikon) or ciphered (Sony).

I'm happy to develop the decryption and document it in my book.  The ciphered code for Sony is discussed here: #906 and is the mostly complex change I every made to Exiv2.  It is discussed in the book.  I'll solve nikon_decrypt() and document/discuss it in the book.

**Not so happy news:**

I'm disappointed that no member of the Community has shown any interest in becoming the maintainer of Exiv2.  I've done everything I can to leave the project in a clean and tidy state and my book explains everything I know about Metadata and Exiv2.  https://clanmills.com/exiv2/book/

I will support, mentor and encourage somebody to maintain Exiv2.  At this moment, it seems unlikely!  However, Exiv2 is a unique and critical resource to the Open Source Graphics Community, and I expect a maintainer will volunteer in 2021.  I intend to give a presentation at LGM in Rennes in May 2021 and to run an afternoon workshop based on my book.

https://discuss.pixls.us/t/maintainer-urgently-needed-for-exiv2/21547

**So now I leave, with my head held high**

I remain optimistic about the future of Exiv2.  Don't think that I will change my mind and return from retirement.  I have to retire some day and that day is ..... **today**.

**Comment by clanmills** [comment]
Success!

```
1022 rmills@rmillsmm-local:~/gnu/exiv2/team/book/build $ ./tvisitor -pR ../files/CH0_0174.NEF | grep LensIDNumber
         187 | 0x00a7 Exif.Nikon.LdLensIDNumber    |     UBYTE |        1 |           | 119
1023 rmills@rmillsmm-local:~/gnu/exiv2/team/book/build $ exiftool -a -G1 -s "-*lens*" ../files/CH0_0174.NEF | grep LensIDNumber
[Nikon]         LensIDNumber                    : 119
1024 rmills@rmillsmm-local:~/gnu/exiv2/team/book/build $ 
```

**Comment by hassec** [comment]
@clanmills should we close this issue, since it's now supported in the maintenance branch?

**Comment by clanmills** [comment]
Yip.  Close it.

We had the meeting with the **darktable** _**(and rawspeed)**_ people on Saturday and we have a plan for v0.27.4.  The meeting notes are on the chat-server and I'll update the Exiv2 RoadMap today.

One day, there will be a 0.28 release from 'master'.  This feature and about 100 PRs will have to merged from 0.27-maintenance into master.  Who will do this and when will that happen?  I have no idea.  If everybody behaved as you have, everything will be fine and a new maintainer will appear to make this happen.
